client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,443 @@
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		self.reserved_only = reserved_only;
+
+		if reserved_only {
+			// Disconnect all non-reserved peers.
+			for (peer_id, state) in self.nodes.iter_mut() {
+				match state {
+					PeerState::Connected(d) => {
+						match d {
+							Direction::Inbound => self.num_in -= 1,
+							Direction::Outbound => self.num_out -= 1,
+						}
+						*state = PeerState::NotConnected;
+						let _ = self.to_notifications.unbounded_send(Message::Drop {
+							set_id: self.set_id,
+							peer_id: *peer_id,
+						});
+					},
+					PeerState::NotConnected => {},
+				}
+			}
+		} else {
+			// Try connecting to non-reserved peers.
+			self.alloc_slots();
+		}
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.nodes.insert(peer_id, PeerState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains(&peer_id) {
+			return
+		}
+
+		match self.nodes.remove(&peer_id) {
+			Some(state) => match state {
+				PeerState::Connected(d) => {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					let _ = self
+						.to_notifications
+						.unbounded_send(Message::Drop { set_id: self.set_id, peer_id });
+				},
+				PeerState::NotConnected => {},
+			},
+			None => {
+				trace!(
+					target: "peerset",
+					"Trying to remove unknown peer {} from {:?}",
+					peer_id, self.set_id,
+				);
+			},
+		}
+	}
+
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		if self.reserved_only && !self.reserved_nodes.contains(&peer_id) {
+			let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+			return
+		}
+
+		// Adding incoming peer to our set of peers even before we decide whether to accept
+		// it is questionable, but this is how it was implemented in the original `Peerset`.
+		let state = self.nodes.entry(peer_id).or_insert(PeerState::NotConnected);
+		match state {
+			// If we're already connected, don't answer, as the docs mention.
+			PeerState::Connected(_) => return,
+			PeerState::NotConnected => {
+				if self.peerset_handle.is_banned(peer_id) {
+					let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+					return
+				}
+
+				let is_no_slot_occupy = self.reserved_nodes.contains(&peer_id);
+				// Note that it is possible for num_in to be strictly superior to the max, in case
+				// we were connected to a reserved node then marked it as not reserved.
+				if self.num_in >= self.max_in && !is_no_slot_occupy {
+					let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+					return
+				}
+
+				if !is_no_slot_occupy {
+					self.num_in += 1;
+				}
```

Comment:
I'm wondering if we could encode the node type into the type system (normal/reserved). It would be less error prone than this practice of increasing and decreasing slots.